In [1]:
# Load Libraries
%matplotlib inline
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats
import pandas as pd 

from sklearn.model_selection import train_test_split
from scipy import stats

from multiprocessing import cpu_count

# Remove Not Informed Variables

Most machine learning models do not allow for missing values in the input data. 

In categorical variables, missing values can be transformed into a new category 'Missing', solving this way the issue.

For numerical variables, these missing values can be filled using different approaches (see notebook 5) but if a variable has too many missing values these approaches are no longer useful and the best approach would be to remove the variable entirely from the dataset.

Let's see here how to select variables with too many missing values and remove them from our dataset.

## Load Data

In [2]:
dat = pd.read_csv("/Users/jesus/Dropbox/Planes/proyectos/academia/IE/2021-2022/ML/dat/i2.csv", sep = ";")
dat.head()

user booking_date origin_airport       price sales channel   ant airline
0  user5   01/11/2018            MAD         NaN        online   NaN      i2
1  user7   01/11/2018            DUB  147.500000        online  38.0      i2
2  user4   02/11/2018            TFS   24.049999        online  19.0      i2
3  user8   29/10/2018            MAD   59.709999        online   8.0      i2
4  user7   01/11/2018            NaN   37.299999   call center   4.0      i2

## Detect Missing Values

Let's count the number of missing values of each variable of the dataset.

In [3]:
dat.apply(lambda x: 100*np.sum(x.isna())/len(x))

user               0.0
booking_date       0.0
origin_airport     0.4
price             57.9
sales channel      0.0
ant                0.5
airline            0.0
dtype: float64

price has 57.9% of missing values! It may be better to remove it from the dataset...

## Detect Non-informed Variables

More generally, we can detect non-informed variables on a dataset doing this:

In [11]:
missing_values_threshold = 50 # You should try different values on a real problem.

In [12]:
non_informed_variables = dat.columns.values[dat.apply(lambda x: 100*np.sum(x.isna())/len(x)) >= missing_values_threshold].tolist()
non_informed_variables

['price']

## Remove Non-informed Variables

Let's use the previously computed list with non-informed column names to remove these variables from our dataset.

In [13]:
new_dat = dat.drop(non_informed_variables, axis = 1)
new_dat

user booking_date origin_airport sales channel   ant airline
0     user5   01/11/2018            MAD        online   NaN      i2
1     user7   01/11/2018            DUB        online  38.0      i2
2     user4   02/11/2018            TFS        online  19.0      i2
3     user8   29/10/2018            MAD        online   8.0      i2
4     user7   01/11/2018            NaN   call center   4.0      i2
..      ...          ...            ...           ...   ...     ...
995   user2   01/11/2018            JMK        online  29.0      i2
996  user10   01/11/2018            SVQ        online  39.0      i2
997   user4   30/10/2018            MAD        online   5.0      i2
998  user10   02/11/2018            CDG        online   4.0      i2
999   user6   01/11/2018            MAD        online   0.0      i2

[1000 rows x 6 columns]

## Define Custom Function

Let's create our own custom function to remove non-informed variables.

In [ ]:
def count_nas(x): 
    ret = 100*np.sum(x.isna())/len(x)
    return ret

In [17]:
def remove_non_informed(X, missing_values_threshold = 50):
    non_informed_variables = X.columns.values[X.apply(lambda x: 100*np.sum(x.isna())/len(x)) >= missing_values_threshold].tolist()
    X = X.drop(non_informed_variables, axis = 1)
    print('Variables ' + str(non_informed_variables) + ' have been removed from dataset.')
    return X;
                                    
                                                                                        

In [18]:
new_dat = remove_non_informed(dat)
new_dat.head()

Variables ['price'] have been removed from dataset.


user booking_date origin_airport sales channel   ant airline
0  user5   01/11/2018            MAD        online   NaN      i2
1  user7   01/11/2018            DUB        online  38.0      i2
2  user4   02/11/2018            TFS        online  19.0      i2
3  user8   29/10/2018            MAD        online   8.0      i2
4  user7   01/11/2018            NaN   call center   4.0      i2

In [19]:
new_dat = remove_non_informed(dat, missing_values_threshold = 60)
new_dat.head()

Variables [] have been removed from dataset.


user booking_date origin_airport       price sales channel   ant airline
0  user5   01/11/2018            MAD         NaN        online   NaN      i2
1  user7   01/11/2018            DUB  147.500000        online  38.0      i2
2  user4   02/11/2018            TFS   24.049999        online  19.0      i2
3  user8   29/10/2018            MAD   59.709999        online   8.0      i2
4  user7   01/11/2018            NaN   37.299999   call center   4.0      i2